cdap-app-fabric/src/main/java/io/cdap/cdap/internal/app/runtime/distributed/remote/RemoteExecutionTwillRunnerService.java

```diff
@@ -617,7 +621,25 @@ public TwillController create(@Nullable Callable<Void> startupTask, long timeout
               } catch (Exception e) {
                 LOG.warn("Force termination of remote process for {} failed", programRunId, e);
               }
-              programStateWriter.error(programRunId, throwable);
+              // Update the state to rejected if failure is due to throttling,
+              // otherwise update to failed.
+              if (throwable instanceof TooManyRequestsException) {
+                Store store = new DefaultStore(txRunner);
+                try {
+                  ProgramDescriptor descriptor = store.loadProgram(programOpts.getProgramId());
+                  programStateWriter.reject(
+                      programRunId, programOpts, descriptor,
+                      Constants.Security.Authentication.RUNTIME_IDENTITY, throwable);
+                } catch (Exception e) {
+                  // Transition to failed state (old behavior) if there are errors
+                  // while transitioning to rejected state.
+                  LOG.warn(
+                    "Error while transitioning {} to 'rejected', transitioning to 'failed' instead.", programRunId, e);
+                  programStateWriter.error(programRunId, throwable);
```

Comment:
Good to add e to suppressed exceptions.

Reply:
Added e as a suppressed exception in throwable.